crates/sui-indexer-alt/src/handlers/coin_balance_buckets.rs

```diff
@@ -0,0 +1,132 @@
+                    let serialized = bcs::to_bytes(&coin_type)
+                        .map_err(|_| anyhow!("Failed to serialize type for {}", object.id()))?;
+
+                    coin_types.insert(object.id(), serialized);
+                }
+            }
+
+            // Deleted and wrapped coins
+            for change in tx.effects.object_changes() {
+                // The object is not deleted/wrapped, or if it is it was unwrapped in the same
+                // transaction.
+                if change.output_digest.is_some() || change.input_version.is_none() {
+                    continue;
+                }
+
+                // Object is not a coin
+                if !coin_types.contains_key(&change.id) {
+                    continue;
+                }
+
+                let object_id = change.id;
+                match values.entry(object_id) {
+                    Entry::Occupied(_) => {}
+
+                    Entry::Vacant(entry) => {
+                        entry.insert(StoredCoinBalanceBucket {
+                            object_id: object_id.to_vec(),
+                            cp_sequence_number: cp_sequence_number as i64,
+                            owner_kind: None,
+                            owner_id: None,
+                            coin_type: None,
+                            coin_balance_bucket: None,
+                        });
+                    }
+                }
+            }
+
+            // Modified and created coins.
+            for object in &tx.output_objects {
+                let object_id = object.id();
+
+                let Some(coin_type) = coin_types.get(&object_id) else {
+                    continue;
+                };
+
+                // Coin balance only tracks address-owned objects
+                let (owner_kind, owner_id) = match object.owner() {
+                    Owner::AddressOwner(owner_id) => (StoredOwnerKind::Address, owner_id),
+                    Owner::Immutable | Owner::ObjectOwner(_) | Owner::Shared { .. } => continue,
+                    Owner::ConsensusV2 { .. } => todo!(),
+                };
+
+                let Some(coin) = object.as_coin_maybe() else {
+                    bail!("Failed to deserialize Coin for {object_id}");
+                };
+
+                match values.entry(object_id) {
+                    Entry::Occupied(_) => {}
+
+                    Entry::Vacant(entry) => {
+                        entry.insert(StoredCoinBalanceBucket {
+                            object_id: object_id.to_vec(),
+                            cp_sequence_number: cp_sequence_number as i64,
+                            owner_kind: Some(owner_kind),
+                            owner_id: Some(owner_id.to_vec()),
+                            coin_type: Some(coin_type.clone()),
+                            coin_balance_bucket: Some((coin.balance.value() as f64).log10() as i16),
```

Comment:
We also need to confirm that the coin was either created in this checkpoint or it was modified and its bucket has changed within this checkpoint, otherwise I'm not sure there's a benefit to using buckets.

Reply:
lol totally forgot